python/ray/data/impl/pandas_block.py

```diff
@@ -0,0 +1,182 @@
+from typing import Dict, List, Union, Tuple, Any, TypeVar, Optional, \
+    TYPE_CHECKING
+
+import collections
+import numpy as np
+
+try:
+    import pandas
+except ImportError:
+    pandas = None
+
+from ray.data.block import Block, BlockAccessor, BlockMetadata
+from ray.data.impl.table_block import TableBlockAccessor, TableRow, \
+    TableBlockBuilder, SortKeyT, GroupKeyT
+from ray.data.impl.arrow_block import ArrowBlockAccessor
+from ray.data.aggregate import AggregateFn
+
+if TYPE_CHECKING:
+    import pyarrow
+    import pandas
+
+T = TypeVar("T")
+
+
+class PandasRow(TableRow):
+    def as_pydict(self) -> dict:
+        # TODO (kfsorm): If the type of column name is not str, e.g. int.
+        # The result keys will stay as int. Should we enforce `str(k)`?
+        # Maybe we don't even allow non-str column names?
+        return {k: v[0] for k, v in self._row.to_dict("list").items()}
+
+    def __getitem__(self, key: str) -> Any:
+        assert isinstance(key, str)
```

Comment:
Arrow column names are always strings, but this doesn't apply to Pandas. e.g.

```python
df = pd.DataFrame({1: np.random.rand(10), 2: np.random.rand(10)})
table = pa.table(df)
```

We can access columns of the above arrow table via `table[0]`, `table[1]`, `table["1"]`, `table["2"]`, but not `table["0"]` or `table[2]`.

We can access the DataFrame object via `df[1]`, `df[2]`, but not `df[0]`.

There are many places that assume that column names of a Ray dataset are strings. Update or redesign is needed.

Reply:
I think it's pretty reasonable to require column names for strings. We can consider pandas block support as an optimization to avoid unnecessary copy/conversions, but adding features that Arrow doesn't support is out of scop.e

Reply:
Column name type check added.

Reply:
Could we raise a better error in this case?